[week3] Spring Web으로 ToDo REST API 만들기

csh9066 opened the pull request:
테스트 네이밍 생각 하느라 너무 고통스러웠네요 ㅜㅜ
테스트 코드 작성하는 좋은 템플릿이나 방법론 같은 팁좀 얻을 수 있을까요!!

csh9066 committed:
TaskController의 성공 케이스만 테스트하는 TaskControllerWebTest를 추가해라

hannut91 commented:
이미 너무 잘하셨는데요 ㅎㅎ given, when, then에 대해서 구분해서 고민해 보시면 좋을 것 같아요. 어떤 것이 주어졌을 때 무엇을 하면 결과가 ~ 나온다. 예를들어 로그인하지 않은 상태로 (Given) 할 일 목록 조회를 시도하면(When) 에러를 응답한다(Then)

https://johngrib.github.io/wiki/junit5-nested/
https://www.betterspecs.org/

csh9066 committed:
TaskControllerWebTest를 DCI 방법으로 테스트를 작성해라

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerWebTest.java:
```diff
@@ -0,0 +1,258 @@
+        }
+
+        @Nested
+        @DisplayName("id가 숫자가 아니면")
```
와 테스트 케이스를 아주 명시적으로 작성해 주셨네요 ㅎㅎ 좋습니다

csh9066 replied:
감사합니다!  할 때 엄청 고통스러웠는데 칭찬들으니까 엄청 보람이 있네요 👍

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerWebTest.java:
```diff
@@ -0,0 +1,258 @@
+
+    private static final String TASKS_DEFAULT_PATH = "/tasks";
+    private static final String TASKS_VALID_ID_PATH = TASKS_DEFAULT_PATH + "/" + TASK_ID;
+    private static final String TASKS_INVALID_ID_PATH = TASKS_DEFAULT_PATH + "/" + INVALID_TASK_ID;
```
아래에서는 존재하지 않는 아이디로 사용하고 있더라구요. 여기서도 같은 문맥으로 드러내는 것은 어떨까요?

hannut91 reviewed (commented):
너무 잘해주셔서 할 말이 없네요 ㅎㅎ좋습니다!

csh9066 committed:
TaskController를 테스트하는 TaskControllerTest를 작성해라

csh9066 committed:
TaskControllerWebTest의 상수 TASK_ID를 좀 더 명확하게 변경해라

csh9066 committed:
TaskControllerTest EOF 추가해라

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerWebTest.java:
```diff
@@ -0,0 +1,257 @@
+            @DisplayName("400 status를 응답한다.")
+            void it_responses_400_status() throws Exception {
+                mockMvc.perform(get(TASKS_NON_NUMBER_ID_PATH))
+                        .andExpect(status().isBadRequest());
```
음.. 다시 보니 여기는 변수 이름으로 작성하는 것보다 오히려, 그냥 문자열로 드러내는게 더 이해가 빠를 수도 있을 것 같아요.

우리가 그냥 두 개의 값을 더하는 `1 + 1`을 `plusToAnother()`로 표현하면 더 이해하기 어려워지는 것처럼요.

```java
                mockMvc.perform(get("/tasks/abcd"))
                        .andExpect(status().isBadRequest());
```

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerWebTest.java:
```diff
@@ -0,0 +1,257 @@
+            @DisplayName("변경된 task를 응답한다.")
+            void it_responses_updated_task() throws Exception {
+                mockMvc.perform(put(TASKS_VALID_ID_PATH)
+                                .content(taskToString(task)) //
```
여기 안쓰이는 주석이 있는 것 같군요

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,179 @@
+    class Describe_list {
+
+        @Test
+        @DisplayName("tasks를 반환한다.")
```
테스트에선 빈 리스트를 반환하고 있네요. task가 있는 경우도 추가해주는 것은 어떨까요?

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,179 @@
+    class Describe_detail {
+
+        @Nested
+        @DisplayName("존재하는 id가 주어지면")
+        class Context_with_exist_id {
+
+            private Long id;
+
+            @BeforeEach
+            void setUp() {
+                Task task = taskController.create(source);
+                id = task.getId();
+            }
+
+            @Test
+            @DisplayName("task를 반환한다.")
+            void it_returns_task() {
+                Task task = taskController.detail(id);
+
+                assertThat(task).isNotNull();
+                assertThat(task.getId()).isEqualTo(id);
+            }
+        }
+
+        @Nested
+        @DisplayName("존재하지 않는 id가 주어지면")
+        class Context_with_non_exist_id {
+
+            @Test
+            @DisplayName("TaskNotFoundException이 발생한다.")
+            void it_throws_task_not_found_exception() {
+                assertThatThrownBy(() -> taskController.detail(TASK_INVALID_ID))
+                        .isInstanceOf(TaskNotFoundException.class);
+            }
+        }
```
위에 테스트에서는 id를 설정하는 곳이 있는데, 아래에는 없네요.

```java
@Nested
@DisplayName("detail 메소드는")
class Describe_detail {
    private Long id;

    @Nested
    @DisplayName("존재하는 id가 주어지면")
    class Context_with_exist_id {

        @BeforeEach
        void setUp() {
            Task task = taskController.create(source);
            id = task.getId();
        }

        @Test
        @DisplayName("task를 반환한다.")
        void it_returns_task() {
            Task task = taskController.detail(id);

            assertThat(task).isNotNull();
            assertThat(task.getId()).isEqualTo(id);
        }
    }

    @Nested
    @DisplayName("존재하지 않는 id가 주어지면")
    class Context_with_non_exist_id {

        @BeforeEach
        void setUp() {
            id = TASK_INVALID_ID;
        }

        @Test
        @DisplayName("TaskNotFoundException이 발생한다.")
        void it_throws_task_not_found_exception() {
            assertThatThrownBy(() -> taskController.detail(id))
                    .isInstanceOf(TaskNotFoundException.class);
        }
  }
}
```

이런식으로 표현하면 `Context`마다 어떻게 달라지는지 보여줄 수 있을거에요. 지금은 물론 좀 과하게 보이긴 하지만요

csh9066 committed:
TaskService를 테스트하는 TaskServiceTest를 작성해라

csh9066 committed:
사용하지 않는 주석을 제거해라

csh9066 committed:
TaskControllerWebTest의 테스트 가독성을 위해 path에 관련된 상수들을 제거해라

csh9066 committed:
TaskServiceTest 잘못된 given을 올바르게 변경해라

updateTask의 존재하는 id가 주어지는 테스트에서 테스트용 task를 생성 하는데 변경되는 source가 아닌 새로운 task가 생성되어야 해서 수정합니다.

csh9066 committed:
TaskControllerTest에서 테스트 가독성을 위해 given의 범위를 좁히도록 변경해라

현재 source라는 변수의 given이 전역에 선언되고 setUp 메소드를 통해 사용하지 않을 때도 생성되고 있습니다. 각 Describe와 Context 마다 사용하는 given만 설정하도록 변경했습니다.

hannut91 commented:
테스트 구조가 아주 잘 잡혀있어서 좋네요! 서비스와 컨트롤러 뿐만 아니라 Entity에 대해서도 테스트를 작성해 보세요

csh9066 committed:
Task를 테스트 하는 TaskTest를 작성해라

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,223 @@
+
+            @BeforeEach
+            void setUp() {
+                Task task = taskService.createTask(givenSource(TASK_TITLE));
```
계속 같은 값으로 `Task`를 만들고 있는 것 같아요. 테스트를 하기 전에 신뢰할 수 있는 테스트용 데이터를 미리 만들어놓고 사용하는 것은 어떨까요?

csh9066 committed:
TaskControllerWebTest에 patch 요청 테스트를 추가해라

csh9066 committed:
TaskControllerTest 테스트에 필요한 given을 다시 전역변수로 올려라

현재 테스트 생성시 필요한 givenSource라는 메소드를 만들어 사용했는데 전역 변수로 올리는게 가독성이 좋다고 판단해 다시 올렸습니다.

csh9066 committed:
TaskServiceTest 테스트에 필요한 given을 전역변수로 올려라

csh9066 committed:
TaskTest를 DCI 형식으로 테스트해라

csh9066 committed:
TaskTest를 getTitle 테스트를 추가해라

csh9066 committed:
TaskControllerWebTest 코드를 정리해라

가독성을 위한 띄어쓰기 추가

hannut91 commented on app/src/test/java/com/codesoom/assignment/models/TaskTest.java:
```diff
@@ -0,0 +1,99 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class TaskTest {
```
`Task` 객체에 대해서 아주 꼼꼼하게 테스트를 작성해 주셨네요! 테스트를 우리가 작성한 코드의 예제라고 생각해 봅시다. 테스트가 꼼꼼하게 작성되어 있다면 하나하나의 기능에 대해서는 좋겠지만, 전체적으로 어떻게 사용하는지는 판단하기 어려울 수도 있을 것 같아요. 따라서 이 객체를 어떻게 사용하는지를 간단하게만 작성해 보는 것도 좋을 것 같아요

```java
Task task = new Task();
task.setId(1L);
task.setTitle("책 읽기");

assertThat(task.getId()).isEqualTo(1L);
assertThat(task.getTitle()).isEqualTo("책 읽기");
```

hannut91 reviewed (approved):
고생 많으셨습니다 ㅎㅎ 과제풀이 영상을 통해 복습을 진행해 주세요

hannut91 commented:
이번 주 과제를 통해서 무엇을 할 수 있게 되셨나요? 주간 회고를 작성해서 공유해 주세요

csh9066 commented:
> 이번 주 과제를 통해서 무엇을 할 수 있게 되셨나요? 주간 회고를 작성해서 공유해 주세요

네 빨리 작성해서 올리겠습니다!
